entry/csp.js

```diff
@@ -0,0 +1,95 @@
+import { createHash } from 'crypto';
+import { parse } from 'node-html-parser';
+import { URL } from 'url';
+
+const scriptTypeIgnoreList = ['application/json', 'application/ld+json'];
+
+const defaultBaseName = 'http://relative-url';
+
+const hashScriptContents = (scriptContents) =>
+  createHash('sha256').update(scriptContents).digest('base64');
+
+export default function createCSPHandler({ extraHosts = [] } = {}) {
+  const hosts = new Set();
+  const shas = new Set();
+
+  const addScriptContents = (contents) => {
+    shas.add(hashScriptContents(contents));
+  };
+
+  const addScriptUrl = (src) => {
+    const { origin } = new URL(src, defaultBaseName);
+
+    if (origin !== defaultBaseName) {
+      hosts.add(origin);
+    }
+  };
+
+  extraHosts.forEach((host) => addScriptUrl(host));
+
+  const processScriptNode = (scriptNode) => {
+    const src = scriptNode.getAttribute('src');
+
+    if (src) {
+      addScriptUrl(src);
+    } else if (
+      !scriptTypeIgnoreList.includes(scriptNode.getAttribute('type'))
+    ) {
+      addScriptContents(scriptNode.firstChild.rawText);
+    }
+  };
+
+  const registerScript = (script) => {
```

Comment:
Is there any way for this script to fail if headers have already been flushed once? If I understand correctly right now the new script will just not be returned in the second head flush, leading to any registredScripts to not be white-listed silently.
```ts
if(csp.hasReturned) {
  throw new Error(`Unable to register script. Content Security Policy already sent. Try registering scripts before calling flushHeadTags. Script: ${script.substr(0, 30)}`);
}
```